Exploration features

danielfdsilva opened the pull request:
This PR adds the features to A&E recently discussed:

- [x] Zarr layer type support
- [x] Modal to get dataset layer tiles (through dataset options menu)
- [x] Show error on non supported datasets for analysis (non raster)
- [x] Fix a couple of nasty bugs
- [x] Add a feature flag to enable the new exploration
   If the site is built with any truthy value set on `FEATURE_NEW_EXPLORATION`, the app shows the page, otherwise all normal business
- [x] Fix the analysis state upon selecting / deselecting AOIs or changing parameters + adding datasets
- [x] Resolve race condition with URL parameters
- [ ] https://github.com/NASA-IMPACT/veda-ui/issues/761
- [ ] https://github.com/NASA-IMPACT/veda-ui/issues/755

danielfdsilva commented:
For now I reenabled the exploration path so you can test it, but you'll need to go there directly (like done so far). We can then completely lock it off.

danielfdsilva commented:
Fix #724

j08lue commented on app/scripts/components/exploration/components/datasets/dataset-options.tsx:
```diff
@@ -30,53 +37,87 @@ export default function DatasetOptions(props: DatasetOptionsProps) {
+                type='button'
+                onClick={() => setTileModalRevealed(true)}
+              >
+                <CollecticonMap title='Open tile url options' /> View tile urls
```
```suggestion
                <CollecticonMap title='Open tile URL options' /> Load into GIS
```

j08lue commented on app/scripts/components/exploration/components/datasets/dataset-options.tsx:
```diff
@@ -30,53 +37,87 @@ export default function DatasetOptions(props: DatasetOptionsProps) {
+          <li>
+            <Tip
+              disabled={!!dataset.meta?.tileUrls}
+              content='Current dataset does not offer tile urls'
```
```suggestion
              content='This dataset does not offer tile URLs'
```

j08lue commented on app/scripts/components/exploration/components/datasets/tile-link-modal.tsx:
```diff
@@ -0,0 +1,98 @@
+      content={
+        <Form>
+          {[
+            { label: 'XYZ Tile Url', value: tileUrls?.xyzTileUrl },
+            { label: 'WMTS Tile Url', value: tileUrls?.wmtsTileUrl }
```
```suggestion
            { label: 'XYZ Tile URL', value: tileUrls?.xyzTileUrl },
            { label: 'WMTS Tile URL', value: tileUrls?.wmtsTileUrl }
```

j08lue reviewed (changes requested):
I tested `/exploration` but just in the preview, so without enabling that feature with the feature flag.

This is what I tested:

1. Open `/exploration`, add CMIP6 (Zarr/Kerchunk/TiTiler-Xarray) dataset -> "Oh no, something went wrong" ❌ 
2. Look up layer tile URLs -> worked nicely, wordsmithy changes suggested below. ✅ (Bonus: A more GIS-like icon could be something like [this](https://www.vectorstock.com/royalty-free-vector/global-export-icon-outline-style-vector-25200438))
3. Tried to run analysis on a vector dataset (EIS `Fire` from mock layers) -> ❌ No error/warning is shown, analysis status looks like success, time series graph area is blank (neither graph nor timeline grey nuggets)

Please let me know whether my testing was wrong or so or you could use more details.

danielfdsilva commented:
1) This is a data problem.
    - Locally I was using the `https://test-stac.delta-backend.com` which returns `summaries: null`. When this happens we default to `extent.temporal.interval` and so it was working.
    - The preview, however was using `https://staging-stac.delta-backend.com` which returns `summaries: { datetime: [null, null] }`, which causes it to break.
    - I added a check for this case, but **we need data guardrails in place**. The app needs the summaries (and other properties) to be able to function, so they need to be provided. If this is not possible, we need to define another way to get the required data - but it has to be consistently available.  

---

3) There was a problem with the error catching which is now fixed. I also added an additional message to inform the user when an analysis did not return data.
<img width="1253" alt="image" src="https://github.com/NASA-IMPACT/veda-ui/assets/1090606/035c12e8-0635-4c73-94d9-cfd017cf2e14">

j08lue reviewed (changes requested):
## Previous review addressed

Vector datasets and the TiTiler-Xarray-served CMIP6 dataset now work fine!


## New issue

Another observation during testing, though: The "Outdated" state of the temporal and spatial selection for analysis does not behave as expected:

A) Moving geometry does not seem to cause "outdated"

1. Add dataset
2. Draw AOI
3. Keep date range as-is
4. Start analysis and wait for completion
5. State:  ✅ up-to-date - correct
6. **Move or change geometry**
7. State: ✅ up-to-date - wrong  (should be outdated)
8. Deselect and select the geometry again
9. State: ❌ outdated - correct

B) Changing date range causes permanent "outdated"

1. Add dataset
2. Draw AOI
3. **Change date range with sliders on timeline**
4. Start analysis and wait for completion
5. State: ❌ outdated - wrong (should be up-to-date)

These are reproducible for me on a clean browser - could still be me somehow...

danielfdsilva commented:
@j08lue You reported the bugs very clearly with perfect reproducibility steps! 🎉 

Just pushed a fix that addressed both things, among other small problems! Let me know if you find anything else!

j08lue reviewed (changes requested):
## Previous review addressed

Awesome, the up-to-date/outdated state now behaves as expected in the situations I tested before.

## New find

When adding datasets to an existing analysis, users do not get an obvious option to run the analysis on those:

1. Run analysis on some dataset(s)
2. State: ✅ up-to-date - correct
3. Add more datasets
4. State: ✅ up-to-date - ?? how would a user now extend the same analysis to the added datasets? Should we use "Outdated" when they added new datasets? Or somehow give them another possibility to add these datasets to the analysis (e.g. on the timeline or something)?

![image](https://github.com/NASA-IMPACT/veda-ui/assets/3404817/f102ce14-2cb7-4c61-b5e1-f9f8580e1a27)

danielfdsilva commented:
@j08lue yet another great find :p
The correct behavior should be that the analysis kicks off automatically. I applied the needed fix.

j08lue reviewed (approved):
All review comments addressed.

Remaining from the PR check list are the new requests
- https://github.com/NASA-IMPACT/veda-ui/issues/761
- https://github.com/NASA-IMPACT/veda-ui/issues/755

Those could also be for a new PR, if you like.
